test/linalg/test_sparse_spmv.fypp

```diff
@@ -0,0 +1,290 @@
+            call dense2coo( dense , COO )
+            
+            allocate( vec_x(5)  , source = 1._wp )
+            allocate( vec_y1(4) , source = 0._wp )
+            allocate( vec_y2(4) , source = 0._wp )
+
+            vec_y1 = matmul( dense, vec_x )
+            
+            call check(error, all(vec_y1 == real([6,11,15,15],kind=wp)) )
+            if (allocated(error)) return
+
+            call spmv( COO, vec_x, vec_y2 )
+            call check(error, all(vec_y1 == vec_y2) )
+            if (allocated(error)) return
+        end block
+        #:endfor
+    end subroutine
+
+    subroutine test_coo2ordered(error)
+        !> Error handling
+        type(error_type), allocatable, intent(out) :: error
+        type(COO_sp) :: COO
+
+        call COO%malloc(4,4,12)
+        COO%data(:) = 1
+        COO%index(:,1) = [1,2]
+        COO%index(:,2) = [1,3]
+        COO%index(:,3) = [1,4]
+        COO%index(:,4) = [2,3]
+        COO%index(:,5) = [2,4]
+        COO%index(:,6) = [3,4]
+
+        COO%index(:,7) = [2,3]
+        COO%index(:,8) = [2,4]
+        COO%index(:,9) = [2,5]
+        COO%index(:,10) = [3,4]
+        COO%index(:,11) = [3,5]
+        COO%index(:,12) = [4,5]
+
+        call coo2ordered(COO,sort_data=.true.)
+        call check(error, COO%nnz < 12 .and. COO%nnz == 9 )
+        if (allocated(error)) return
+
+        call check(error, all(COO%data==[1,1,1,2,2,1,2,1,1])  )
+        if (allocated(error)) return
+
+        call check(error, all(COO%index(1,:)==[1,1,1,2,2,2,3,3,4])  )
+        if (allocated(error)) return
+
+        call check(error, all(COO%index(2,:)==[2,3,4,3,4,5,4,5,5])  )
+        if (allocated(error)) return
+
+    end subroutine 
+
+    subroutine test_csr(error)
+        !> Error handling
+        type(error_type), allocatable, intent(out) :: error
+        #:for k1, t1, s1 in (KINDS_TYPES)
+        block
+            integer, parameter :: wp = ${k1}$
+            type(CSR_${s1}$) :: CSR
+            ${t1}$, allocatable :: vec_x(:)
+            ${t1}$, allocatable :: vec_y(:)
+
+            call CSR%malloc(4,5,10)
+            CSR%data(:)   = real([9,-3,4,7,8,-1,8,4,5,6],kind=wp)
+            CSR%col(:)    = [1,5,1,2,2,3,4,1,3,4]
+            CSR%rowptr(:) = [1,3,5,8,11]
+            
+            allocate( vec_x(5) , source = 1._wp )
+            allocate( vec_y(4) , source = 0._wp )
+            call spmv( CSR, vec_x, vec_y )
+            
+            call check(error, all(vec_y == real([6,11,15,15],kind=wp)) )
+            if (allocated(error)) return
+        end block
+        #:endfor
+    end subroutine
+
+    subroutine test_csc(error)
+        !> Error handling
+        type(error_type), allocatable, intent(out) :: error
+        #:for k1, t1, s1 in (KINDS_TYPES)
+        block
+            integer, parameter :: wp = ${k1}$
+            type(CSC_${s1}$) :: CSC
+            ${t1}$, allocatable :: vec_x(:)
+            ${t1}$, allocatable :: vec_y(:)
+
+            call CSC%malloc(4,5,10)
```

Comment:
As I think I've commented earlier, I would really love to see a unified and consistent matrix setting interface for all matrices that would take in 3 arrays: `i, j, v` containing row, column, and value of each non-zero value. 

I think the internals should be left `public` for the advanced user, but this simple API would be very useful to reduce potential errors. For example: automatically remove zeros, ensure correct storage pointers, etc.